### PR TITLE
Support direct cell message

### PR DESCRIPTION
**Pull request by yanchengnv**
### Description

This PR is intended to boost overall performance of the simulator. It does so in two areas:
- Use in-memory calls for comms between two cells within the same process. This is 2X faster than even local tcp.
- Reduce number of unnecessary threads. Instead of creating them at component init time, create them only on as-needed basis. This strategy is applied to fed_event and cellnet. In case of fed event, the "poster" thread is created only when fed events are received. Most jobs don't use fed events.  In case of cellnet, the threads for handling queued messages are also made on as-needed basis. Since we do not use the queue message mechanism in FLARE, this will save two extra threads per cell.  All these will have a huge saving on number of threads when simulating large number of clients.

The removing of fed event's "poster" thread will also reduce the overall shutdown time.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/fuel/f3/cellnet/cell.py [hunk 1]
```diff
@@ -505,10 +527,13 @@ def is_backbone_ready(self):
 
     def _set_bb_for_client_root(self):
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py:528]: backbone

File: nvflare/fuel/f3/cellnet/cell.py [hunk 2]
```diff
@@ -1238,18 +1293,26 @@ def fire_and_forget(
...
         if isinstance(targets, str):
             targets = [targets]
 
         message.set_header(MessageHeaderKey.OPTIONAL, optional)
         with self.bulk_lock:
+            if self.bulk_checker is None:
+                self.logger.info(f"{self.my_info.fqcn}: starting bulk_checker")
+                self.bulk_checker = threading.Thread(target=self._check_bulk)
+                self.bulk_checker.start()
+                self.logger.info(f"{self.my_info.fqcn}: started bulk_checker")
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py:1308]: check bulk messages. Bulk message is a message that contains many other messages that are queued. Currently FLARE is not using this mechanism yet.

File: nvflare/fuel/f3/cellnet/cell.py [hunk 3]
```diff
@@ -175,6 +176,10 @@ def log_messaging_error(logger, log_text: str, cell, msg: Union[Message, None],
         logger.debug(log_text)
         if log_except:
             logger.debug(secure_format_traceback())
+    elif log_warn:
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py:179]: Sounds like a good idea. Will make the change.

File: nvflare/widgets/fed_event.py [hunk 4]
```diff
@@ -104,6 +102,11 @@ def _receive(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareab
             return make_reply(ReturnCode.BAD_REQUEST_DATA)
 
         with self.in_lock:
+            if self.poster is None:
+                # create the poster thread now
+                self.poster = threading.Thread(target=self._post, args=())
+                self.poster.start()
```
Comment (YuanTingHsieh) [inline nvflare/widgets/fed_event.py:108]: should we change the line in `__init__.py` to `self.poster = None`

File: nvflare/widgets/fed_event.py [hunk 5]
```diff
@@ -104,6 +102,11 @@ def _receive(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareab
             return make_reply(ReturnCode.BAD_REQUEST_DATA)
 
         with self.in_lock:
+            if self.poster is None:
+                # create the poster thread now
+                self.poster = threading.Thread(target=self._post, args=())
```
Comment (nvidianz) [inline nvflare/widgets/fed_event.py:107]:
```suggestion
                self.poster = threading.Thread(target=self._post, name="poster", args=())
```

File: nvflare/fuel/f3/cellnet/cell.py [hunk 6]
```diff
@@ -1238,18 +1293,26 @@ def fire_and_forget(
...
+            raise RuntimeError(f"{self.get_fqcn()}: bulk message is not enabled!")
+
         if isinstance(targets, str):
             targets = [targets]
 
         message.set_header(MessageHeaderKey.OPTIONAL, optional)
         with self.bulk_lock:
+            if self.bulk_checker is None:
+                self.logger.info(f"{self.my_info.fqcn}: starting bulk_checker")
+                self.bulk_checker = threading.Thread(target=self._check_bulk)
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell.py:1306]:
```suggestion
                self.bulk_checker = threading.Thread(target=self._check_bulk, name="bulk_checker“)
```

File: nvflare/fuel/f3/cellnet/cell.py [hunk 7]
```diff
@@ -1273,6 +1336,11 @@ def _receive_bulk_message(self, request: Message):
         target_msgs = request.payload
         assert isinstance(target_msgs, list)
         with self.bulk_msg_lock:
+            if self.bulk_processor is None:
+                self.logger.debug(f"{self.my_info.fqcn}: starting bulk message processor")
+                self.bulk_processor = threading.Thread(target=self._process_bulk_messages)
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell.py:1341]:
```suggestion
                self.bulk_processor = threading.Thread(target=self._process_bulk_messages， name="bulk_proc")
```

**Review by nvidianz — APPROVED** [review]

File: nvflare/fuel/f3/cellnet/cell.py [hunk 8]
```diff
@@ -160,7 +161,7 @@ def __init__(self, targets: List[str]):
         self.received_replies = {}
 
 
-def log_messaging_error(logger, log_text: str, cell, msg: Union[Message, None], log_except=False):
+def log_messaging_error(logger, log_text: str, cell, msg: Union[Message, None], log_except=False, log_warn=False):
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py:164]: good idead

File: nvflare/fuel/f3/cellnet/cell.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py]: Changed to use a simpler solution: logger.log(log_level...).

File: nvflare/widgets/fed_event.py (hunk 4 above)
Comment (yanchengnv) [reply nvflare/widgets/fed_event.py]: Changed as suggested.

File: nvflare/fuel/f3/cellnet/cell.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py]: Changed as suggested.

File: nvflare/widgets/fed_event.py (hunk 5 above)
Comment (yanchengnv) [reply nvflare/widgets/fed_event.py]: Changed as suggested.

File: nvflare/fuel/f3/cellnet/cell.py (hunk 7 above)
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py]: Changed as suggested

File: nvflare/fuel/f3/cellnet/cell.py (hunk 6 above)
Comment (yanchengnv) [reply nvflare/fuel/f3/cellnet/cell.py]: Changed as suggested

File: nvflare/fuel/f3/cellnet/cell.py [hunk 9]
```diff
@@ -1837,14 +1904,30 @@ def get_sub_cell_names(self) -> Tuple[List[str], List[str]]:
 
         Returns: fqcns of child cells, fqcns of top-level client cells
         """
-        children = []
-        clients = []
+        children_dict = {}
+        clients_dict = {}
```
Comment (YuanTingHsieh) [inline nvflare/fuel/f3/cellnet/cell.py:1908]: just a note, we could also use set here since we only care about keys.

**Review by YuanTingHsieh — APPROVED** [review]
LGTM

**Pull request merged**